pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/SegmentStateSnapshot.java

```diff
@@ -0,0 +1,123 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.broker.routing.instanceselector;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+import org.apache.pinot.common.metrics.BrokerMeter;
+import org.apache.pinot.common.metrics.BrokerMetrics;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ *  This class represents a snapshot state of segments used for routing purpose.
+ *  Note that this class is immutable after creation.
+ *
+ *  For old segments, we return a list of online instances with online flags set to true.
+ *  For old segments without any online instances, we report them as unavailable segments.
+ *
+ *  For new segments, we return a list of candidate instance with online flags to indicate whether the instance is
+ *  online or not.
+ *  We don't report new segment as unavailable segments because it is valid for new segments to be not online at all.
+ */
+@Immutable
+public class SegmentStateSnapshot {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentStateSnapshot.class);
+
+  private Map<String, List<SegmentInstanceCandidate>> _segmentCandidates;
+  private Set<String> _unavailableSegments;
+
+  // Create a segment state snapshot based on some in-memory states to be used for routing.
```

Comment:
(nit) We also use `/** ... **/` for the comments on the functions.

Reply:
done